File: vacuum/emWave.f
      program emWave
      implicit none 
      integer xn, yn, zn, tn 
      parameter(xn=2001,tn=3001)
      real*8 ds,dt
      parameter(ds=1d-2,dt=1d-3)
      real*8 s(xn),t(tn)
      real*8 e(xn,3), b(xn,3)
      real*8 eo(xn,3), bo(xn,3)
      integer i,j,k,m
      real*8 pi 
      parameter(pi=3.1415927d0)
      open(1,file='emWave_data.js',status='unknown')
c    set environment 
      do i=1,xn       
        s(i)=-10+(i-1)*1d-2
      end do
      do m=1,tn
        t(m)=(m-1)*1d-2
      end do 
c    initial condition 
      do i=1,xn
        if(abs(s(i))<=3.0)then
          eo(i,2)= dsin(s(i)*pi)*(dcos(s(i)*pi/3)+1d0)/2d0
          eo(i,3)= dcos(s(i)*pi)*(dcos(s(i)*pi/3)+1d0)/2d0
c          eo(i,2)= (dcos(s(i)*pi)+1d0)/2d0
c          eo(i,3)= (dcos(s(i)*pi)+1d0)/2d0
          bo(i,2)=-dcos(s(i)*pi)*(dcos(s(i)*pi/3)+1d0)/2d0
          bo(i,3)=dsin(s(i)*pi)*(dcos(s(i)*pi/3)+1d0)/2d0
        else 
          eo(i,2)=0d0
          bo(i,3)=0d0
        end if
c        eo(i,2)=0d0
c        eo(i,3)=0d0
        bo(i,2)=0d0
        bo(i,3)=0d0
      end do 
      
      write(1,*)'var s=['
      write(1,'("{x:",f8.3,", y:0., z:0.},")')(s(i),i=1,xn-1,10)
      write(1,'("{x:",f8.3,", y:0., z:0.}]")')s(xn)
      write(1,*)'var e=new Array()'
      write(1,*)'var b=new Array()'
c    start 
      do m=1,tn-100
        do i=1+10,xn-10
          b(i,2)=bo(i,2)+dt*(eo(i+1,3)-eo(i-1,3))/(2d0*ds)
          b(i,3)=bo(i,3)-dt*(eo(i+1,2)-eo(i-1,2))/(2d0*ds)
          e(i,2)=eo(i,2)-dt*(bo(i+1,3)-bo(i-1,3))/(2d0*ds)
          e(i,3)=eo(i,3)+dt*(bo(i+1,2)-bo(i-1,2))/(2d0*ds)
        end do 
c        if(m==2)then
        if(mod(m,20)==1)then
          write(1,*)'e[',m/20,']=['
          write(1,'("{x:",f8.3,",y:",f8.3,",z:",f8.3,"},")')
     |         (e(i,1),e(i,2),e(i,3),i=1,xn-1,10)
          write(1,'("{x:",f8.3,",y:",f8.3,",z:",f8.3,"}]")')
     |         e(xn,1),e(xn,2),e(xn,3)

          write(1,*)'b[',m/20,']=['
          write(1,'("{x:",f8.3,",y:",f8.3,",z:",f8.3,"},")')
     |         (b(i,1),b(i,2),b(i,3),i=1,xn-1,10)
          write(1,'("{x:",f8.3,",y:",f8.3,",z:",f8.3,"}]")')
     |         b(xn,1),b(xn,2),b(xn,3)

        end if
        do i=1,xn
          eo(i,2)=e(i,2)
          eo(i,3)=e(i,3)
          bo(i,2)=b(i,2)
          bo(i,3)=b(i,3)
        end do 
      end do 
c      write(1,'(1x,f8.3,1x,f8.3,1x,f8.3)')(s(i),e(i,2),b(i,3),i=1,xn,10)
c      write(1,*)'}'
      stop
      end 

File: vacuum/emWave.js
var fs=require('fs');
var xn=2001,yn=1,zn=1,tn=1001
var dx=0.01,dt=0.01
var s=new Array(10,10),t=new Array()
var eo=new Array(), bo=new Array()
var e =new Array(), b =new Array()
var i,j,k,m
var pi=Math.PI
// set envionment
for(i=1;i<xn;i++){
  s[i]=-10.+(i-1)*dx
}
for(m=1;m<tn;m++){
  t[m]=(m-1)*dt
}
// initial condition
eo[2]=new Array; eo[3]=new Array;
bo[2]=new Array; bo[3]=new Array;
for(i=1;i<=xn;i++){
  if(Math.abs(s[i])<=1.){
    eo[2][i]=Math.cos(s[i]*pi)+1 
  }else{
    eo[2][i]=0
  }
  eo[3][i]=0
  bo[2][i]=0; bo[3][i]=0
}
fs.writeFile('./emWave.json',JSON.stringify(eo,null,4),function(){})

// start 
e[2]=new Array; e[3]=new Array;
b[2]=new Array; b[3]=new Array;
for(m=1;m<tn-100;m++){
  for(i=1+10;i<=xn-10;i++){
    e[2][i]=eo[2][i]+dt*(bo[3][i+1]-bo[3][i-1])/(2*dx)
    e[3][i]=eo[3][i]-dt*(bo[2][i+1]-bo[2][i-1])/(2*dx)
    b[2][i]=bo[2][i]-dt*(eo[3][i+1]-eo[3][i-1])/(2*dx)
    b[3][i]=bo[3][i]+dt*(eo[2][i+1]-eo[2][i-1])/(2*dx)
  }
  for(i=1;i<=xn;i++){
    eo[2][i]=e[2][i]
    eo[3][i]=e[3][i]
    bo[2][i]=b[2][i]
    bo[3][i]=b[3][i]
  }
  if(m==400){
    console.log(e[2][11],e[2][12],e[2][13])
    for(i=1;i<=xn;i+=10){
      console.log(i,e[2][i])
    }
  }
}

